File: fireplace-app/src/main/java/io/github/bric3/fireplace/charts/XYDataset.java
package io.github.bric3.fireplace.charts;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * A dataset containing zero, one or many (x, y) data items.
 */
public class XYDataset {
    @NotNull
    public final List<XY<Long, Double>> items;
    @NotNull
    public final Range<Long> rangeOfX;
    @NotNull
    public final Range<Double> rangeOfY;
    @NotNull
    public final String label;

    /**
     * Creates a new dataset.
     *
     * @param sourceItems the source items ({@code null} not permitted).
     * @param label The label for the dataset.
     */
    public XYDataset(@NotNull List<@NotNull XY<@NotNull Long, @NotNull Double>> sourceItems, @NotNull String label) {
        this.label = label;
        // verify that none of the x-values is null or NaN ot INF
        // and while doing that record the mins and maxes
        Objects.requireNonNull(sourceItems, "sourceItems must not be null");
        Objects.requireNonNull(label, "label must not be null");
        if (sourceItems.isEmpty()) {
            throw new IllegalArgumentException("sourceItems must not be empty");
        }
        this.items = new ArrayList<>(sourceItems);
        long minX = Long.MAX_VALUE;
        long maxX = Long.MIN_VALUE;
        double minY = Double.POSITIVE_INFINITY;
        double maxY = Double.NEGATIVE_INFINITY;
        for (var sourceItem : sourceItems) {
            Objects.requireNonNull(sourceItem, "elements of sourceItems must not be null");
            minX = Math.min(minX, sourceItem.x);
            maxX = Math.max(maxX, sourceItem.x);

            minY = Math.min(minY, sourceItem.y);
            maxY = Math.max(maxY, sourceItem.y);
        }
        this.rangeOfX = new Range<>(minX, maxX);
        this.rangeOfY = tweakYRange(new Range<>(minY, maxY));
    }

    protected Range<Double> tweakYRange(Range<Double> yRange) {
        return yRange;
    }

    public int getItemCount() {
        return items.size();
    }

    public Long xAt(int index) {
        return this.items.get(index).x;
    }

    public Double yAt(int index) {
        return this.items.get(index).y;
    }

    /**
     * Compare with another instance.
     * Note doesn't compare the data points, only the ranges and the label.
     *
     * @param o the object to compare with this instance.
     * @return whether two data set are assumed to be equal.
     */
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        XYDataset xyDataset = (XYDataset) o;
        return rangeOfX.equals(xyDataset.rangeOfX) && rangeOfY.equals(xyDataset.rangeOfY) && label.equals(xyDataset.label);
    }

    @Override
    public int hashCode() {
        return Objects.hash(rangeOfX, rangeOfY, label);
    }

    /**
     * A pair of objects.
     *
     * @param <X> the type of the first object.
     * @param <Y> the type of the second object.
     */
    public record XY<X extends Number, Y extends Number>(X x, Y y) {}
}

File: fireplace-app/src/main/kotlin/io/github/bric3/fireplace/charts/XYDataset.kt
/*
 * Fireplace
 *
 * Copyright (c) 2021, Today - Brice Dutheil
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package io.github.bric3.fireplace.charts

import java.util.*
import kotlin.math.max
import kotlin.math.min

/**
 * A dataset containing zero, one or many (x, y) data items.
 *
 * @param sourceItems the source items (`null` not permitted).
 * @param label The label for the dataset.
 */
open class XYDataset(
    sourceItems: List<XY<Long, Double>>,
    final override val label: String
) : ChartDataset {
    private val items: List<XY<Long, Double>>
    final override val rangeOfX: Range<Long>
    final override val rangeOfY: Range<Double>

    init {
        // verify that none of the x-values is null or NaN ot INF
        // and while doing that record the mins and maxes
        Objects.requireNonNull(sourceItems, "sourceItems must not be null")
        Objects.requireNonNull(label, "label must not be null")
        require(sourceItems.isNotEmpty()) { "sourceItems must not be empty" }
        this.items = ArrayList(sourceItems)
        var minX = Long.MAX_VALUE
        var maxX = Long.MIN_VALUE
        var minY = Double.POSITIVE_INFINITY
        var maxY = Double.NEGATIVE_INFINITY
        for (sourceItem in sourceItems) {
            Objects.requireNonNull(sourceItem, "elements of sourceItems must not be null")
            minX = min(minX.toDouble(), sourceItem.x.toDouble()).toLong()
            maxX = max(maxX.toDouble(), sourceItem.x.toDouble()).toLong()

            minY = min(minY, sourceItem.y)
            maxY = max(maxY, sourceItem.y)
        }
        this.rangeOfX = Range(minX, maxX)
        @Suppress("LeakingThis")
        this.rangeOfY = tweakYRange(Range(minY, maxY))
    }

    protected open fun tweakYRange(yRange: Range<Double>): Range<Double> {
        return yRange
    }

    override val itemCount: Int
        get() = items.size

    override fun xAt(index: Int): Long {
        return items[index].x
    }

    override fun yAt(index: Int): Double {
        return items[index].y
    }

    /**
     * Compare with another instance, ignoring data points.
     *
     * Only compare the ranges and the label.
     *
     * @param other the object to compare with this instance.
     * @return whether two data set are assumed to be equal.
     */
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val xyDataset = other as XYDataset
        return rangeOfX == xyDataset.rangeOfX && rangeOfY == xyDataset.rangeOfY && label == xyDataset.label
    }

    /**
     * Hash this instance, ignoring data points.
     *
     * Only hashes the ranges and the label.
     * @return The hash.
     */
    override fun hashCode(): Int {
        return Objects.hash(rangeOfX, rangeOfY, label)
    }

    /**
     * A pair of objects.
     *
     * @param <X> the type of the first object.
     * @param <Y> the type of the second object.
     */
    @JvmRecord
    data class XY<X : Number, Y : Number>(val x: X, val y: Y)
}
